File: client/account.py
import utils
import requests

def account():
    utils.clear_screen()
    username = utils.read_config("username")
    password = utils.read_config("password")
    formatted_password = f"{password[0:3]}{('*' * (len(password) - 5) if len(password) > 5 else '')}{password[-2:]}" if len(password) >= 5 else password
    print("Account setup")
    print(f"Current username: {username}")
    print(f"Current password: {formatted_password}")
    if input("Do you want to change any details? (y/n) ").lower() == "y":
        new_username, new_password = username, password
        while new_username == username:
            new_username = input("New username: (Enter to preserve) ")
        while new_password == password:
            new_password = input("New password: (Enter to preserve) ")

        address = f"{utils.read_config('server_url')}/online/change_data"
        if new_username != username and new_password != password:
            mode = 'everything'
        elif new_username != username:
            mode = 'username'
        elif new_password != password:
            mode = 'password'
        else:
            print("Nothing to change")
            mode = ""

        response = None
        response2 = None
        if mode in ['everything', "username"]:
            response = requests.get(f'{address}/user?user={username}&auth={password}&new_user={new_username}')
        if mode in ['everything', "password"]:
            target_user = username if mode == 'password' else new_username
            response2 = requests.get(f"{address}/auth?user={target_user}&auth={password}&new_auth={new_password}")

        if response is not None and response2 is not None:
            if response.status_code == 200 and response2.status_code == 200:
                print("Changed the username and password successfully")
            elif response.status_code == 200:
                print("Changed the username successfully")
            elif response2.status_code == 200:
                print("Changed the password successfully")
            else:
                print(f'Failed to change the account details: {response.text}, {response2.text}')
            if response.status_code != 200 and mode in ["everything", "username"]:
                print(f"Failed to change the username: {response.text}")
            if response2.status_code != 200 and mode in ["everything", "password"]:
                print(f"Failed to change the password: {response2.text}")
            utils.write_config("password", new_password)
            utils.write_config("username", new_username)
    if input("Do you want to delete your account? (y/n) ").lower() == "y":
        if input("Confirm by typing 'delete' ").lower() == "delete":
            address = f"{utils.read_config('server_url')}/online/delete_account"
            response = requests.get(f"{address}?user={username}&auth={password}")
            if response.status_code == 200:
                print("Deleted the account successfully")
                utils.write_config("password", "")
                utils.write_config("username", "")
            else:
                print(f'Failed to delete the account: {response.text}')
    input('Press `Enter` to continue...')

File: flutter/lib/account.dart
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'package:wordle/utils.dart';
import 'package:provider/provider.dart';

class AccountDetailsPage extends StatefulWidget {
  const AccountDetailsPage({super.key});

  @override
  State<AccountDetailsPage> createState() => _AccountDetailsPageState();
}

class _AccountDetailsPageState extends State<AccountDetailsPage> {
  String _username = '';
  String _password = '';
  String _newUsername = '';
  String _newPassword = '';
  String _serverUrl = '';
  String _result = '';
  String _resultDeletion = '';
  late TextEditingController _usernameController;
  late TextEditingController _passwordController;

  @override
  void initState() {
    super.initState();
    _loadUsername();
    _loadPassword();
    _loadServerUrl();
    _usernameController = TextEditingController();
    _passwordController = TextEditingController();
  }

  @override
  void dispose() {
    _usernameController.dispose();
    _passwordController.dispose();
    super.dispose();
  }

  Future<void> _loadUsername() async {
    final username = await getConfig("username");
    setState(() {
      _username = username ?? '';
      _newUsername = _username;
      _usernameController.text = _newUsername;
    });
  }

  Future<void> _loadPassword() async {
    final password = await getConfig("password");
    setState(() {
      _password = password ?? '';
      _newPassword = _password;
      _passwordController.text = _newPassword;
    });
  }

  Future<void> _loadServerUrl() async {
    final serverUrl = await getConfig("server_url");
    setState(() {
      _serverUrl = serverUrl!;
    });
  }

  Future<String> applyChanges() async {
    try {
      final url = '$_serverUrl/online/change_data';
      bool usernameChanged = _username != _newUsername;
      bool passwordChanged = _password != _newPassword;

      if (!usernameChanged && !passwordChanged) {
        return 'Nothing to change.';
      }

      http.Response? usernameResponse;
      http.Response? passwordResponse;

      if (usernameChanged) {
        usernameResponse = await http.get(Uri.parse('$url/user?user=$_username&auth=$_password&new_user=$_newUsername')).timeout(const Duration(seconds: 7));
      }

      if (passwordChanged) {
        final user = usernameChanged ? _newUsername : _username;
        passwordResponse = await http.get(Uri.parse('$url/auth?user=$user&auth=$_password&new_auth=$_newPassword')).timeout(const Duration(seconds: 7));
      }

      bool usernameSuccess = usernameResponse?.statusCode == 200;
      bool passwordSuccess = passwordResponse?.statusCode == 200;

      if (usernameChanged && passwordChanged) {
        if (usernameSuccess && passwordSuccess) {
          return 'Changed the username and password successfully';
        } else if (usernameSuccess) {
          return 'Changed the username successfully, but failed to change password: ${passwordResponse?.body}';
        } else if (passwordSuccess) {
          return 'Changed the password successfully, but failed to change username: ${usernameResponse?.body}';
        } else {
          return 'Failed to change the account details: ${usernameResponse?.body} ${passwordResponse?.body}';
        }
      } else if (usernameChanged) {
        return usernameSuccess ? 'Changed the username successfully' : 'Failed to change the username: ${usernameResponse?.body}';
      } else {
        return passwordSuccess ? 'Changed the password successfully' : 'Failed to change the password: ${passwordResponse?.body}';
      }
    } catch (e) {
      return 'Failed to change the account details: $e';
    }
  }

  Future<String> deleteAccount() async {
    try {
      final url = '$_serverUrl/online/delete_account';
      final response = await http.get(Uri.parse('$url?user=$_username&auth=$_password')).timeout(const Duration(seconds: 7));

      if (response.statusCode == 200) {
        return 'Account deleted successfully.';
      } else {
        return 'Failed to delete the account: ${response.body}';
      }
    } catch (e) {
      return 'Failed to delete the account: $e';
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: const Text('Account Details'),
      ),
      body: SingleChildScrollView(
        child: Column(
          children: [
            ListTile(
              title: const Text('Current username'),
              trailing: SizedBox(
                width: 200,
                child: TextField(
                  readOnly: true,
                  controller: TextEditingController(text: _username),
                )
              )
            ),
            ListTile(
              title: const Text('Current password'),
              trailing: SizedBox(
                width: 200,
                  child: TextField(
                    readOnly: true,
                    controller: TextEditingController(text: '•'*_password.length),
                  )
              )
            ),
            ListTile(
              title: const Text('New username'),
              trailing: SizedBox(
                width: 200,
                child: TextField(
                  controller: _usernameController,
                  decoration: InputDecoration(
                    hintText: _newUsername.isNotEmpty ? _newUsername : 'Enter your new username',
                  ),
                  onChanged: (value) async {
                    setState(() {
                      _newUsername = value;
                    });
                    final connProvider = Provider.of<ConnectionStateProvider>(context, listen: false);
                    final accProvider = Provider.of<AccountStateProvider>(context, listen: false);
                    await setConfig("username", value);
                    connProvider.forceCheck();
                    accProvider.forceCheck();
                  },
                ),
              )
            ),
            ListTile(
              title: const Text('New password'),
              trailing: SizedBox(
                width: 200,
                child: TextField(
                  controller: _passwordController,
                  decoration: InputDecoration(
                    hintText: _newPassword.isNotEmpty ? '•'*_newPassword.length : 'Enter your new password',
                  ),
                  obscureText: true,
                  onChanged: (value) async {
                    setState(() {
                      _newPassword = value;
                    });
                    final connProvider = Provider.of<ConnectionStateProvider>(context, listen: false);
                    final accProvider = Provider.of<AccountStateProvider>(context, listen: false);
                    await setConfig("password", value);
                    connProvider.forceCheck();
                    accProvider.forceCheck();
                  },
                ),
              )
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: ElevatedButton(
                    onPressed: () async {
                      String result = await applyChanges();
                      setState(() {
                        _result = result;
                        if (result.contains('Changed')) {
                          _password = _newPassword;
                          _username = _newUsername;
                        }
                      });
                      if (_result.contains('Failed to change the account details.')) {
                        showErrorToast(_result, long: true);
                      }
                    },
                    child: Text('Apply')
                  ),
                ),
                Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: ElevatedButton(
                    onPressed: () async {
                      String resultDeletion = await deleteAccount();
                      setState(() {
                        _resultDeletion = resultDeletion;
                        if (resultDeletion == "Account deleted successfully.") {
                          _password = '';
                          _username = '';
                        }
                      });
                      if (_result.contains('Failed')) {
                        showErrorToast(_resultDeletion, long: true);
                      }
                    },
                    child: Text('Delete the account')
                  ),
                ),
              ],
            ),
            if (_result != '')
              Padding(
                padding: const EdgeInsets.all(8.0),
                child: Text(
                  _result,
                  style: TextStyle(fontSize: 16, color: (_result.contains('Failed to change the account details.') ? Colors.red : Theme.of(context).colorScheme.onSurface)),
                ),
              ),
            if (_resultDeletion != '')
              Padding(
                padding: const EdgeInsets.all(8.0),
                child: Text(
                  _resultDeletion,
                  style: TextStyle(fontSize: 16, color: (_resultDeletion.contains('Failed to delete the account') ? Colors.red : Theme.of(context).colorScheme.onSurface)),
                ),
              ),
          ],
        ),
      ),
    );
  }
}